pkg/sql/sem/tree/constant.go

```diff
@@ -545,10 +554,29 @@ func (expr *StrVal) ResolveAsType(ctx *SemaContext, typ *types.T) (Datum, error)
 		return &expr.resString, nil
 	case types.BytesFamily:
 		return ParseDByte(expr.s)
-	}
 
-	datum, err := ParseAndRequireString(typ, expr.s, ctx)
-	return datum, err
+	case types.ArrayFamily,
+		types.DateFamily,
+		types.TimeFamily,
+		types.TimeTZFamily,
+		types.TimestampFamily,
+		types.TimestampTZFamily:
+		// Interpreting a string as one of these types may depend on the timezone or
+		// the current time; if we do it now, the result won't be safe to reuse
+		// later. So in this case we return a CastExpr and let the conversion happen
+		// at evaluation time. We still want to error out if the conversion is not
+		// possible though.
+		_, err := ParseAndRequireString(typ, expr.s, semaCtx)
```

Comment:
can we re-use the cast volatility information here to be more exact?

Reply:
Good call, done.